### PR TITLE
catalog: remove outdated CHECK post-deserialization changes

### LINKED ISSUE #120017: sql: ALTER that drops and re-adds a CHECK constraint leaves table in an invalid state
**Describe the problem**
Run the following commands:
```
CREATE TABLE t (
  a INT PRIMARY KEY,
  b TEXT,
  CHECK (b IN ('x', 'y'))
);

ALTER TABLE t
DROP CONSTRAINT check_b,
ADD CONSTRAINT check_b CHECK (b IN ('x', 'y', 'z'));
```

Then you can see that there is a leftover mutation in the descriptor:
```
> select crdb_internal.pb_to_json('cockroach.sql.sqlbase.Descriptor', descriptor) from system.descriptor where id = 't'::regclass::oid;
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      crdb_internal.pb_to_json
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  {"table": {"checks": [{"columnIds": [2], "constraintId": 3, "expr": "b IN ('x':::STRING, 'y':::STRING, 'z':::STRING)", "name": "check_b", "validity": "Validating"}], "columns": [{"id": 1, "name": "a", "type": {"family": "IntFamily", "oid": 20, "width": 64}}, {"id": 2, "name": "b", "nullable": true, "type": {"family": "StringFamily", "oid": 25}}], "createAsOfTime": {"wallTime": "1709759323199984000"}, "families": [{"columnIds": [1, 2], "columnNames": ["a", "b"], "defaultColumnId": 2, "name": "primary"}], "formatVersion": 3, "id": 104, "modificationTime": {}, "mutations": [{"constraint": {"check": {"columnIds": [2], "constraintId": 3, "expr": "b IN ('x':::STRING, 'y':::STRING, 'z':::STRING)", "name": "crdb_internal_constraint_3_name_placeholder", "validity": "Validating"}, "foreignKey": {}, "name": "crdb_internal_constraint_3_name_placeholder", "uniqueWithoutIndexConstraint": {}}, "direction": "ADD", "mutationId": 1, "state": "WRITE_ONLY"}], "name": "t", "nextColumnId": 3, "nextConstraintId": 4, "nextFamilyId": 1, "nextIndexId": 2, "nextMutationId": 1, "parentId": 100, "primaryIndex": {"constraintId": 1, "createdAtNanos": "1709759323199965000", "encodingType": 1, "foreignKey": {}, "geoConfig": {}, "id": 1, "interleave": {}, "keyColumnDirections": ["ASC"], "keyColumnIds": [1], "keyColumnNames": ["a"], "name": "t_pkey", "partitioning": {}, "sharded": {}, "storeColumnIds": [2], "storeColumnNames": ["b"], "unique": true, "version": 4}, "privileges": {"ownerProto": "root", "users": [{"privileges": "2", "userProto": "admin", "withGrantOption": "2"}, {"privileges": "2", "userProto": "root", "withGrantOption": "2"}], "version": 3}, "replacementOf": {"time": {}}, "unexposedParentSchemaId": 101, "version": "3"}}
```
Specifically:
```
"mutations": [{"constraint": {"check": {"columnIds": [2], "constraintId": 3, "expr": "b IN ('x':::STRING, 'y':::STRING, 'z':::STRING)", "name": "crdb_internal_constraint_3_name_placeholder", "validity": "Validating"}, "foreignKey": {}, "name": "crdb_internal_constraint_3_name_placeholder", "uniqueWithoutIndexConstraint": {}}, "direction": "ADD", "mutationId": 1, "state": "WRITE_ONLY"}]
```

In v23.1 and earlier, this does not occur.


**Expected behavior**
No leftover mutations

Epic CRDB-35306

Jira issue: CRDB-36434

### PR DESCRIPTION
**Pull request by rafiss**
This post-deserialization changes is vestigial. It is no longer
needed since during the upgrade to 23.2, all descriptors had changes
applied.

The maybeAddConstraintIDs change had a bug that caused it to clobber the
mutation of an in-progress schema change. Luckily we can just remove the
change entirely rather than fixing it.

fixes https://github.com/cockroachdb/cockroach/issues/120017

Release note (bug fix): Fixed a bug that occurred when using ALTER TABLE
to drop and re-add a CHECK constraint with the same name.

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/alter_table [hunk 1]
```diff
@@ -3598,3 +3598,34 @@ CREATE UNIQUE INDEX ON public.t_118626(c) WHERE n>= 1;
 
 statement ok
 ALTER TABLE public.t_118626 ALTER PRIMARY KEY USING COLUMNS(b);
+
+# Regression test to make sure that we can drop and re-add a CHECK constraint
```
Comment (annrpom) [inline pkg/sql/logictest/testdata/logic_test/alter_table:3602]: Looks like the legacy schema changer doesn't like this test - are we able to specify "do this only in dsc land" in logic tests?

**Review by annrpom — APPROVED** [review]
lgtm mod some test rewrites - the tests that are failing make sense as well (i.e. we don't seem to be breaking anything we shouldn't; every test failing seems to have an assumption that we don't meet anymore)

File: pkg/sql/logictest/testdata/logic_test/alter_table (hunk 1 above)
Comment (rafiss) [reply pkg/sql/logictest/testdata/logic_test/alter_table:3602]: there is! i added a skip directive here for the legacy schema changer

**Comment by rafiss** [comment]
tftr!

bors r+

**Pull request merged**